### PR TITLE
Notify the embedder when it should display or hide an IME

### PR DESCRIPTION
**Pull request by fabricedesre**
This adds a couple of embedder messages triggered when an editable element is focused or blured. The embedder also gets the type of data to edit so it can display a different keyboard type or a custom input method eg. for color choosing.

This is a partial fix for issue #12127 

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach build-geckolib` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).

- [ ] There are tests for these changes OR
- [X] These changes do not require tests because there are no tests for the embedding api :(

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/20676)

### CODE REVIEW

File: components/constellation/constellation.rs [hunk 1]
```diff
@@ -1247,6 +1247,14 @@ impl<Message, LTF, STF> Constellation<Message, LTF, STF>
             FromScriptMsg::SetFullscreenState(state) => {
                 self.embedder_proxy.send(EmbedderMsg::SetFullscreenState(source_top_ctx_id, state));
             }
+            FromScriptMsg::ShowIME(kind) => {
+                debug!("constellation got ShowIME message");
+                self.embedder_proxy.send(EmbedderMsg::ShowIME(source_top_ctx_id, kind));
```
Comment (cbrewster) [inline components/constellation/constellation.rs]: I think I would rather have us talk to the embedder from the script thread instead of going through the constellation.

Comment (fabricedesre) [reply components/constellation/constellation.rs]: I didn't know this was a thing since we don't do it for other cases like TitleChange.

Comment (cbrewster) [reply components/constellation/constellation.rs]: IRC Chat: https://mozilla.logbot.info/servo/20180421#c14649458-c14649484
TL;DR: We should avoid adding lots of cases where the constellation is only forwarding a message when the message could be sent directly. But, I would be okay if this were a followup that removed other instances of this.

Also in the case of `TitleChange`, the constellation is checking if the source pipeline was the top-level pipeline.

File: components/script/dom/element.rs [hunk 2]
```diff
@@ -1087,6 +1088,22 @@ impl Element {
...
 
+    // Returns the kind of IME control needed for a focusable element, if any.
+    pub fn input_method_type(&self) -> Option<InputMethodType> {
+        if !self.is_focusable_area() {
+            return None;
+        }
+
+        if let Some(input) = self.downcast::<HTMLInputElement>() {
+            input.input_type().as_ime_type()
+        } else if let Some(_textarea) = self.downcast::<HTMLTextAreaElement>() {
```
Comment (cbrewster) [inline components/script/dom/element.rs]: nit: `else if self.is::<HTMLTextAreaElement>() {`

File: components/script_traits/script_msg.rs [hunk 3]
```diff
@@ -163,6 +163,10 @@ pub enum ScriptMsg {
     GetScreenSize(IpcSender<(DeviceUintSize)>),
     /// Get the available screen size (pixel)
     GetScreenAvailSize(IpcSender<(DeviceUintSize)>),
+    /// Request to present an IME to the user when an editable element is focused.
+    ShowIME(InputMethodType),
+    /// Request to hide the IME when the editable element is blured.
```
Comment (cbrewster) [inline components/script_traits/script_msg.rs]: nit: s/blured/blurred/

**Review by cbrewster — APPROVED** [review]

**Comment by fabricedesre** [comment]
@cbrewster Is there anything left for me to do here?

**Comment by cbrewster** [comment]
LGTM @bors-servo r+
